storage/innobase/trx/trx0undo.cc

```diff
@@ -975,13 +984,20 @@ trx_undo_mem_create_at_db_start(trx_rseg_t *rseg, ulint id, uint32_t page_no)
 	ut_ad(id < TRX_RSEG_N_SLOTS);
 
 	mtr.start();
-	const buf_block_t* block = buf_page_get(
-		page_id_t(rseg->space->id, page_no), 0, RW_X_LATCH, &mtr);
+	page_id_t page_id{rseg->space->id, 0};
+	/* Access the tablespace header to recover space->free_limit. */
+	const buf_block_t* block = buf_page_get(page_id, 0, RW_X_LATCH, &mtr);
```

Comment:
Before I added this extra load of page 0, yes, the `space->free_limit` would be initialized as 320 in `fil_node_t::read_page0()`. In a number of tests, the undo tablespace (I think it was the system tablespace) was incorrectly found as corrupted, when trying to access undo page 320. Requesting the page via the buffer pool will force the redo log to be applied. In the failing test case that I debugged, there had not been any log checkpoint since the shutdown during server bootstrap.

Reply:
I realize that when merging into 11.0 this code may have to be adjusted for 850d61736deba354377634cf344256ee850b93b7. No worries, this should be highlighted by recovery test failures.

Reply:
We’d better access that page already in `trx_rseg_array_init()`, so that we can validate against `space->free_limit` also there.